presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java

```diff
@@ -84,4 +91,16 @@ public NodeMap(
     {
         return allNodesByHostAndPort;
     }
+
+    public NodeProvider getActiveNodeProvider(NodeSelectionHashStrategy nodeSelectionHashStrategy)
+    {
+        switch (nodeSelectionHashStrategy) {
+            case MODULAR_HASHING:
+                return new ModularHashingNodeProvider(activeNodes);
+            case CONSISTENT_HASHING:
+                return new ConsistentHashingNodeProvider(activeNodesByConsistentHashing);
+            default:
+                throw new IllegalArgumentException(format("Unknown NodeSelectionHashStrategy: %s", nodeSelectionHashStrategy));
+        }
+    }
```

Comment:
Why do we create one for each query? we can move the creation into `NodeMap` constructor, and `getActiveNodeProvider` only return the corresponding instance

In this way, we can move the specific node logic into `ConsistentHashingNodeProvider` (the one we were discussing in NodeScheduler.java)

Reply:
Sure, but this constructor is really cheap. It doesn't really matter.